### PR TITLE
feat(smart-contracts): UP/UDT swap contract

### PR DESCRIPTION
**Pull request by clemsos**
# Description

This is a contract to swap UDT for UP.

Unlock Labs is planning to submit a proposal to the DAO to "move" to Base. This includes deploying a new Governor Contract, a new Timelock contract, as well as a new governance token contract, called UPToken. (Unlock Protocol Token).
However, we want this new token to be a "replacement" for the legacy UDT contract, which means that the only way to "receive" UP tokens is to bridge UDT to Base and then send the UDT tokens to a Swap contract. The amount of UP tokens received is based on a fixed redeem rate of 1000 for 1.

We also want the contract to support the "opposite" operation that would let anyone who owns UP tokens redeem them for legacy UDT, as the inverse rate of 1000 for 1.

# Issues

Fixes #
Refs #

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [x] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [x] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

**Comment by clemsos** [comment]
relates to #13875

File: smart-contracts/contracts/tokens/UP/UPSwap.sol [hunk 1]
```diff
@@ -0,0 +1,158 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.21;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts-upgradeable5/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable5/proxy/utils/Initializable.sol";
+import {IAllowanceTransfer} from "../../interfaces/IAllowanceTransfer.sol";
+
+contract UPSwap is Initializable, OwnableUpgradeable {
+  // 1 UDT for 100 UP
```
Comment (clemsos) [inline smart-contracts/contracts/tokens/UP/UPSwap.sol:10]:
```suggestion
  // 1 UDT for 1000 UP
```

File: smart-contracts/contracts/tokens/UP/UPSwap.sol [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+    // send the UDT to recipient
+    uint amountUDT = amountUP / RATE;
+
+    bool UDTSent = udt.transfer(recipient, amountUDT);
+    if (!UDTSent) {
+      revert TransferFailed(address(udt));
+    }
+    emit UPSwapped(spender, amountUDT, amountUP, recipient);
+  }
```
Comment (clemsos) [inline smart-contracts/contracts/tokens/UP/UPSwap.sol:157]:
```suggestion
```

File: smart-contracts/contracts/tokens/UP/UPSwap.sol [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    // check balance
+    if (udt.balanceOf(spender) < amountUDT) {
+      revert BalanceTooLow(address(udt), spender, amountUDT);
+    }
+
+    // check allowance
+    if (udt.allowance(spender, address(this)) < amountUDT) {
+      revert AllowanceTooLow();
+    }
+
```
Comment (clemsos) [inline smart-contracts/contracts/tokens/UP/UPSwap.sol:77]:
```suggestion
```

**Comment by clemsos** [comment]
### Tokens on Base Sepolia:

- UDT: https://sepolia.etherscan.io/address/0x447B1492C5038203f1927eB2a374F5Fcdc25999d 
- UPToken: https://sepolia.basescan.org/address/0xc9F29DdBD4D828cFb2EB491E9d48013a9c0E3C89

### Try the workflow 

1. Bridge UDT to Base Sepolia

- UDT on Sepolia: https://sepolia.etherscan.io/address/0x447B1492C5038203f1927eB2a374F5Fcdc25999d 
- Bridge UDT: https://superbridge.app/base-sepolia 

2. Swap UDT for UP token 

- Approve Swap to spend UDT on Base Sepolia: https://sepolia.basescan.org/token/0x68a8011d72E6D41bf7CE9dC49De0aeaEBAAC9b39#writeContract
- Swap UDT for Up: https://sepolia.basescan.org/address/0xaB82D702A4e0cD165072C005dc504A21c019718F

example tx: https://sepolia.basescan.org/tx/0x024c51dc08912ed0403f9dc26b3f94e9dc760c0ce64fa9ce68acfe41f19fcdb3

**Comment by julien51** [comment]
ok @bytes032, this is the PR in which we're adding the contract to be audited. Please bear with us as we're cleaning this up!

**Comment by julien51** [comment]
@clemsos please confirm we need to rebase after https://github.com/unlock-protocol/unlock/pull/13875 has been merged?

**Comment by clemsos** [comment]
> @clemsos please confirm we need to rebase after https://github.com/unlock-protocol/unlock/pull/13875 has been merged?

Yes I just merged and will rebase

**Comment by julien51** [comment]
Hey @marvinkruse please check this PR regarding the audit!

**Comment by marvinkruse** [comment]
Hey @julien51 - thank you for tagging me, checking it out!

**Review by julien51 — APPROVED** [review]

**Pull request merged**